### PR TITLE
ARROW-13444: [C++] Remove usage of deprecated std::result_of

### PR DESCRIPTION
**Pull request by bkietz**
Also clean up some UBSAN nullptr arithmetic warnings

### CODE REVIEW

File: cpp/src/arrow/util/future.h [hunk 1]
```diff
@@ -48,7 +49,19 @@ template <typename T>
 struct is_future<Future<T>> : std::true_type {};
 
 template <typename Signature>
-using result_of_t = typename std::result_of<Signature>::type;
+using result_of_t_2 = typename std::result_of<Signature>::type;
```
Comment (pitrou) [inline cpp/src/arrow/util/future.h:52]: Is this actually used?

**Review by pitrou — COMMENTED** [review]
Just one question.

File: cpp/src/arrow/util/future.h (hunk 1 above)
Comment (bkietz) [reply cpp/src/arrow/util/future.h:52]: No, sorry- I was using this during debugging. I'll remove it

**Comment by bkietz** [comment]
@pitrou anything else?

**Review by pitrou — APPROVED** [review]

**Review by westonpace — APPROVED** [review]
LGTM.  Can I ask, for my own learning, why you would choose to use decltype instead of `std::result_of`?

Also, thanks for taking care of the null pointer arithmetic.  I'll close my PR in favor of this one.

**Comment by anders-wind** [comment]
> 
> 
> LGTM. Can I ask, for my own learning, why you would choose to use decltype instead of `std::result_of`?

`std::result_of` is deprecated in c++17 and removed in c++20. See https://issues.apache.org/jira/browse/ARROW-13444

**Comment by westonpace** [comment]
> > LGTM. Can I ask, for my own learning, why you would choose to use decltype instead of `std::result_of`?
> 
> `std::result_of` is deprecated in c++17 and removed in c++20. See https://issues.apache.org/jira/browse/ARROW-13444

Ah, yes, thank you.  I should have been more clear and said `arrow::detail::result_of` but looking closer I see that utility is in `future.h` and not in a more generally accessible utility class.

**Pull request closed without merging**